File: mycode/Grids.m
global agrid bgrid ygrid dagrid dbgrid ymarkovdiag ymarkovoff

% productivity process
if ngpy==1
	logygrid = 0;
	ygrid = 1;
	ymarkov = 0;
	ytrans = 1;
	ydist = 1;
elseif ReadEarningsProcess
    fid1 = fopen(fullfile(EarningsProcessDir,'ygrid_combined.txt'),'r'); logygrid = fscanf(fid1,'%f'); fclose(fid1);
	fid1 = fopen(fullfile(EarningsProcessDir,'ydist_combined.txt'),'r'); ydist = fscanf(fid1,'%f'); fclose(fid1);
    fid1 = fopen(fullfile(EarningsProcessDir,'ymarkov_combined.txt'),'r'); ymarkov = fscanf(fid1,'%f'); fclose(fid1);
	if AdjustProdGridFrisch; logygrid = logygrid/(1+adjfricshgridfrac*frisch); end
	ygrid = exp(logygrid);
    ymarkov = reshape(ymarkov,ngpy,ngpy)'; % since fortran reads in column major order
    ymarkov = ymarkov-diag(sum(ymarkov,2)); % fix up rounding in markov matrix
    ydist = ydist/sum(ydist); % fix up rounding in ergodic distribution
elseif TwoPointWageProcess
	ygrid = [0.8 1.2]';
	logygrid = log(ygrid);
	ytrans = [1-0.06667 0.06667; 0.06667 1-0.06667];
	ydist = [0.5 0.5]';
	ymarkov = ytrans-eye(ngpy); % assumes ytrans is quarterly	
end

ymarkovdiag = diag(ymarkov);
ymarkovoff = ymarkov-diag(ymarkovdiag);
ymarkovdiag = shiftdim(ymarkovdiag,-2);

% adjust mean productivity
lmean = ydist'*ygrid;
ygrid = shiftdim(meanlabeff*ygrid/lmean,-2);
ydist = shiftdim(ydist,-2);

agrid = PowerSpacedGrid(ngpa,agridparam,0,amax)';

% with low gridparam points get bunched close to zero, so evenly space first 8 points;
if ngpa>10
	agrid(1:9) = (0:8)*agrid(10)/9;
end
dagrid = diff(agrid);

% liquid grid
if Borrowing
	nbl = -lumptransfer/(rborr+PerfectAnnuityMarkets*deathrate);
 	abl = max(nbl+cmin,blim);
    if Display>1
        fprintf('natural borrowing limit = %.15g\n',nbl);
        fprintf('actual borrowing limit = %.15g\n',abl);
    end
    bgrid = PowerSpacedGrid(ngpbNEG/2+1,bgridparamNEG,abl,abl/2);
	bgrid = [bgrid abl-fliplr(bgrid(2:end-1)) PowerSpacedGrid(ngpbPOS,bgridparam,0,bmax)];
else
	bgrid = PowerSpacedGrid(ngpb,bgridparam,0,bmax);
end
dbgrid = diff(bgrid);

diff2 = @(X) (X([2:end end])-X([1 1:end-1]))/2;
adelta = diff2(agrid);
bdelta = diff2(bgrid);
abdelta = adelta*bdelta;
ABdelta = abdelta(:)'./abdelta(:);
ABdelta = [[diag(ABdelta,-1);0] ...
           [0;diag(ABdelta,1)] ...
           [diag(ABdelta,-ngpa);zeros(ngpa,1)] ...
           [zeros(ngpa,1);diag(ABdelta,ngpa)]];

File: Fortran/Grids.f90
SUBROUTINE Grids

USE Parameters
USE Globals
USE Procedures

IMPLICIT NONE

INTEGER			:: ia,ib,iy,iaby,iab
REAL(8)         :: lmean
REAL(8)         :: leye(ngpy,ngpy)

REAL(8), EXTERNAL :: golden

!identity matrix
leye = 0.0; DO iy = 1,ngpy
	leye(iy,iy) = 1.0
END DO

!productivity process
IF(ngpy==1) THEN
	logygrid = 0.0
	ygrid = 1.0
	ymarkov = 0.0
	ytrans = 1.0
	ydist = 1.0

ELSE IF (ReadEarningsProcess==1) THEN
	OPEN(1, FILE = trim(EarningsProcessDir) // '/ygrid_combined.txt');READ(1,*) logygrid;CLOSE(1)
	OPEN(1, FILE = trim(EarningsProcessDir) // '/ydist_combined.txt');READ(1,*) ydist;CLOSE(1)
	OPEN(1, FILE = trim(EarningsProcessDir) // '/ymarkov_combined.txt');READ(1,*) ymarkov;CLOSE(1)
	IF (AdjustProdGridFrisch==1) logygrid = logygrid/ (1.0+adjfricshgridfrac*frisch)
	ygrid = exp(logygrid)
	ymarkov = TRANSPOSE(ymarkov) !since fortran reads in column major order	
	!fix up rounding in markov matrix
	DO iy = 1,ngpy
		ymarkov(iy,iy) = ymarkov(iy,iy) - SUM(ymarkov(iy,:))
	END DO
	
ELSE IF (TwoPointWageProcess==1) THEN
	ygrid(1) = 0.8
	ygrid(2) = 1.2
	logygrid = log(ygrid)
	
	ytrans(1,1) = 1.0-0.06667
	ytrans(1,2) = 0.06667
	ytrans(2,1) = 0.06667
	ytrans(2,2) = 1.0-0.06667
	ydist(1) = 0.5
	ydist(2) = 0.5
	ymarkov = (ytrans-leye)/1.0 !assumes ytrans is quarterly	
END IF

ymarkovdiag = 0.0
DO iy = 1,ngpy
	ymarkovdiag(iy,iy) = ymarkov(iy,iy)
END DO
ymarkovoff = ymarkov-ymarkovdiag

!adjust mean productivity
lmean = DOT_PRODUCT(ydist,ygrid)
ygrid = meanlabeff*ygrid/lmean

CALL PowerSpacedGrid (ngpa,agridparam,0.0_8,amax,agrid)

!with low gridparam points get bunched close to zero, so evenly space first 8 points;
IF(ngpa>10) THEN
	DO ia = 1,10-1
		agrid(ia) = (ia-1)*agrid(10)/(10.0-1.0)
	END DO
END IF
dagrid = agrid(2:ngpa)-agrid(1:ngpa-1)

!liquid grid
IF(Borrowing==0) THEN
	CALL PowerSpacedGrid (ngpb,bgridparam,0.0_8,bmax,bgrid)
ELSE IF(Borrowing==1) THEN
	CALL PowerSpacedGrid (ngpbPOS,bgridparam,0.0_8,bmax,bgrid(ngpbNEG+1:ngpb))
	nbl = -lumptransfer/(rborr+PerfectAnnuityMarkets*deathrate)
 	abl = max(nbl + cmin,blim)
	IF (Display>=2) write(*,*) 'natural borrowing limit = ',nbl
	IF (Display>=2) write(*,*) 'actual borrowing limit = ',abl
	CALL PowerSpacedGrid (ngpbNEG/2+1,bgridparamNEG,abl,(abl+bgrid(ngpbNEG+1))/2.0,bgrid(1:ngpbNEG/2+1))
	DO ib = ngpbNEG/2+2,ngpbNEG
		bgrid(ib) = bgrid(ngpbNEG+1) -(bgrid(ngpbNEG+2-ib)-bgrid(1))
	END DO
END IF
dbgrid = bgrid(2:ngpb)-bgrid(1:ngpb-1)


adelta(1) = 0.5*dagrid(1)
adelta(2:ngpa-1) = 0.5*(dagrid(1:ngpa-2)+dagrid(2:ngpa-1))
adelta(ngpa) = 0.5*dagrid(ngpa-1)

bdelta(1) = 0.5*dbgrid(1)
bdelta(2:ngpb-1) = 0.5*(dbgrid(1:ngpb-2)+dbgrid(2:ngpb-1))
bdelta(ngpb) = 0.5*dbgrid(ngpb-1)


!combined grids for openmp loops
iab = 1
DO ia = 1,ngpa
DO ib = 1,ngpb
	afromab(iab) = ia
	bfromab(iab) = ib
	abfromab(ia,ib) = iab
	abdelta(iab) = adelta(ia)*bdelta(ib)
	iab = iab+1
END DO
END DO

iaby = 1
DO ia = 1,ngpa
DO ib = 1,ngpb
DO iy = 1,ngpy
	afromaby(iaby) = ia
	bfromaby(iaby) = ib
	yfromaby(iaby) = iy
	abyfromaby(ia,ib,iy) = iaby
	abydelta(iaby) = adelta(ia)*bdelta(ib)
	abfromaby(iaby) = abfromab(ia,ib)
	iaby = iaby+1
END DO
END DO
END DO


END SUBROUTINE Grids

